tests/jdom/button.spec.ts

```diff
@@ -0,0 +1,31 @@
+// Test the ButtonServer, intended as a test for the unit test framework.
+// Yes, we're meta-testing!
+
+import { suite, test } from "mocha"
+import { ButtonEvent } from "../../src/jdom/constants"
+import { withBus, JDBusTestUtil } from "./tester"
+import { assert } from "../../src/jdom/utils"
+import ButtonServer from "../../src/servers/buttonserver"
+
+suite("button server", () => {
+    test("fires edge events after changing state", async function () {
+        const buttonServer = new ButtonServer("button", false)
+
+        await withBus([{ server: buttonServer }], async (bus, serviceMap) => {
+            const busTest = new JDBusTestUtil(bus) // TODO needs better name, also boilerplate?
+            const buttonService = serviceMap.get(buttonServer) // TODO boilerplate, think about how to eliminate
+
+            buttonServer.down()
+            assert(
+                (await busTest.nextEventWithin(buttonService, { within: 100 }))
+                    .code == ButtonEvent.Down
+            )
+
+            buttonServer.up()
+            assert(
+                (await busTest.nextEventWithin(buttonService, { within: 100 }))
+                    .code == ButtonEvent.Up
+            )
+        })
+    })
+})
```

Comment:
can you add more tests?

Reply:
Added a test for button hold, which I think is the only other interesting functionality in ButtonServer?

If you mean more devices, I think the plan here is to use the VM / adapters project to drive test cases for this. But if you have thoughts other servers that would be good to test I can probably write some.

Reply:
You can also test repeated holds. Probably good enough of a start.

Reply:
Done with two holds (the test is a bit long at 1.7s, because bus time = wall time, so I probably don't want to go more)

Reply:
Button should raise a hold event every 500ms until up is detected.... anothe test.

Btw are these tests run as part of the build?

Reply:
Long tests will motivate better scheduling for unit testing, so keep them.

Reply:
Added a test for repeated hold events.

These tests are run as part of `yarn test` in jacdac-ts.

Long tests: will do! Actually, there's a long test warning for every test case here!
```
  button server
    ✔ fires edge events after changing state (620ms)
    ✔ fires both down and hold events when held (1080ms)
    ✔ repeatedly raise hold events when held (2665ms)
    ✔ detects repeated holds (1714ms)
```